br/pkg/lightning/restore/check_info.go

```diff
@@ -66,252 +46,69 @@ func (rc *Controller) isSourceInLocal() bool {
-}
-
-// clusterResource check cluster has enough resource to import data. this test can by skipped.
-func (rc *Controller) clusterResource(ctx context.Context, localSource int64) error {
-	passed := true
-	message := "Cluster resources are rich for this import task"
-	defer func() {
-		rc.checkTemplate.Collect(Critical, passed, message)
-	}()
-
-	var (
-		clusterAvail  uint64
-		clusterSource uint64
-	)
-	if rc.taskMgr == nil {
-		var err error
-		clusterAvail, err = rc.getClusterAvail(ctx)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		clusterSource = uint64(localSource)
-	} else {
-		if err := rc.taskMgr.CheckTasksExclusively(ctx, func(tasks []taskMeta) ([]taskMeta, error) {
-			clusterAvail = 0
-			clusterSource = 0
-			restoreStarted := false
-			for _, task := range tasks {
-				if task.status > taskMetaStatusInitial {
-					restoreStarted = true
-				}
-				clusterSource += task.sourceBytes
-				if task.clusterAvail > 0 {
-					clusterAvail = task.clusterAvail
-				}
-			}
-			if restoreStarted || clusterAvail > 0 {
-				return nil, nil
-			}
-
-			var err error
-			clusterAvail, err = rc.getClusterAvail(ctx)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-			newTasks := append([]taskMeta(nil), tasks...)
-			for i := 0; i < len(newTasks); i++ {
-				newTasks[i].clusterAvail = clusterAvail
-			}
-			return newTasks, nil
-		}); err != nil {
-			return errors.Trace(err)
-		}
+		return errors.Trace(err)
 	}
-
-	replicaCount, err := rc.getReplicaCount(ctx)
+	result, err := theChecker.Check(checkCtx)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	estimateSize := clusterSource * replicaCount
-	if estimateSize > clusterAvail {
-		passed = false
-		message = fmt.Sprintf("Cluster doesn't have enough space, available is %s, but we need %s",
-			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
-	} else {
-		message = fmt.Sprintf("Cluster available is rich, available is %s, we need %s",
-			units.BytesSize(float64(clusterAvail)), units.BytesSize(float64(estimateSize)))
+	if result != nil {
+		rc.checkTemplate.Collect(result.Severity, result.Passed, result.Message)
 	}
 	return nil
 }
 
 // ClusterIsAvailable check cluster is available to import data. this test can be skipped.
-func (rc *Controller) ClusterIsAvailable(ctx context.Context) {
-	passed := true
-	message := "Cluster is available"
-	defer func() {
-		rc.checkTemplate.Collect(Critical, passed, message)
-	}()
-	checkCtx := WithPreInfoGetterDBMetas(ctx, rc.dbMetas)
-	if err := rc.preInfoGetter.CheckVersionRequirements(checkCtx); err != nil {
-		err = common.NormalizeError(err)
-		passed = false
-		message = fmt.Sprintf("cluster available check failed: %s", err.Error())
-	}
-}
-
-func isTiFlash(store *pdtypes.MetaStore) bool {
-	for _, label := range store.Labels {
-		if label.Key == "engine" && label.Value == "tiflash" {
-			return true
-		}
-	}
-	return false
-}
-
-func (rc *Controller) checkEmptyRegion(ctx context.Context) error {
```

Comment:
Extracted as `(*emptyRegionCheckItem).Check()`, moved to `br/pkg/lightning/restore/precheck_impl.go`, L194